app/src/main/java/com/example/weather/MainActivity.java

```diff
@@ -82,27 +89,28 @@ private void onClickByCity(View view) {
             public void run() {
                 Log.d(TAG, "run started");
 
-                String respond;
-                String temperature;
                 try {
-                    respond = doRequest(request);
-                    if (TextUtils.isEmpty(respond)) {
-                        notificationOnError(getText(R.string.error_wrong_request).toString());
-                        return;
-                    }
-
-                    temperature = parseTemperature(respond);
-                    if (TextUtils.isEmpty(temperature)) {
-                        notificationOnError(getText(R.string.error_wrong_request).toString());
-                        return;
-                    }
-                } catch (IOException | JSONException e) {
+                    doRequest(request, new RequestCallback() {
+                        @Override
+                        public void onRequestSucceed(String respond) {
+                            try {
+                                String temperature = parseTemperature(respond);
```

Comment:
Let parseTemperature return Optional<String> and move try {} catch inside the method.